cfdm/data/subsampledarray.py

```diff
@@ -0,0 +1,1274 @@
+    interpolation method.
+
+    See CF section 8.3 "Lossy Compression by Coordinate Subsampling"
+    and Appendix J "Coordinate Interpolation Methods".
+
+    >>> tie_point_indices={{package}}.TiePointIndex(data=[0, 4, 7, 8, 11])
+    >>> w = {{package}}.InterpolationParameter(data=[5, 10, 5])
+    >>> coords = {{package}}.SubsampledArray(
+    ...     interpolation_name='quadratic',
+    ...     compressed_array={{package}}.Data([15, 135, 225, 255, 345]),
+    ...     shape=(12,),
+    ...     tie_point_indices={0: tie_point_indices},
+    ...     parameters={"w": w},
+    ...     parameter_dimensions={"w": (0,)},
+    ... )
+    >>> print(coords[...])
+    [ 15.          48.75        80.         108.75       135.
+     173.88888889 203.88888889 225.         255.         289.44444444
+     319.44444444 345.        ]
+
+    **Cell boundaries**
+
+    When the tie points array represents bounds tie points then the
+    *shape* parameter describes the uncompressed bounds shape. See CF
+    section 8.3.9 "Interpolation of Cell Boundaries".
+
+    >>> bounds = {{package}}.SubsampledArray(
+    ...     interpolation_name='quadratic',
+    ...     compressed_array={{package}}.Data([0, 150, 240, 240, 360]),
+    ...     shape=(12, 2),
+    ...     tie_point_indices={0: tie_point_indices},
+    ...     parameters={"w": w},
+    ...     parameter_dimensions={"w": (0,)},
+    ... )
+    >>> print(bounds[...])
+    [[0.0 33.2]
+     [33.2 64.8]
+     [64.8 94.80000000000001]
+     [94.80000000000001 123.2]
+     [123.2 150.0]
+     [150.0 188.88888888888889]
+     [188.88888888888889 218.88888888888889]
+     [218.88888888888889 240.0]
+     [240.0 273.75]
+     [273.75 305.0]
+     [305.0 333.75]
+     [333.75 360.0]]
+
+    .. versionadded:: (cfdm) 1.9.TODO.0
+
+    """
+
+    _flag_names = ("location_use_3d_cartesian",)
+
+    def __new__(cls, *args, **kwargs):
+        """Store subarray classes.
+
+        If a child class requires different subarray classes than the
+        ones defined here, then they must be defined in the __new__
+        method of the child class.
+
+        .. versionadded:: (cfdm) 1.9.TODO.0
+
+        """
+        instance = super().__new__(cls)
+
+        instance._Subarray = {
+            "bi_linear": BiLinearSubarray,
+            "bi_quadratic_latitude_longitude": BiQuadraticLatitudeLongitudeSubarray,
+            "linear": LinearSubarray,
+            "quadratic": QuadraticSubarray,
+            "quadratic_latitude_longitude": QuadraticLatitudeLongitudeSubarray,
+        }
+
+        return instance
+
+    def __init__(
+        self,
+        interpolation_name=None,
+        compressed_array=None,
+        shape=None,
+        computational_precision=None,
+        interpolation_description=None,
+        tie_point_indices={},
+        parameters={},
+        parameter_dimensions={},
+        dependent_tie_points={},
+        dependent_tie_point_dimensions={},
```

Comment:
See previous comment regarding the danger of mutable default arguments.

Reply:
Fixed in https://github.com/NCAS-CMS/cfdm/pull/168/commits/069b197504f6c293a818d4996c1ff47093850059